### PR TITLE
Use the new offset type and implement new queries of Akka 2.4.12 

### PR DESCRIPTION
**Pull request by johanandren**
Refs #138

I'm not quite sure about removing the deprecated implementations and calling the new ones wrt extra allocations for every event etc vs avoiding duplication. 

Please let me know if you think I should add it back @patriknw

### CODE REVIEW

File: src/main/scala/akka/persistence/cassandra/query/scaladsl/CassandraReadJournal.scala [hunk 1]
```diff
@@ -249,6 +252,89 @@ class CassandraReadJournal(system: ExtendedActorSystem, config: Config)
...
+   * You can retrieve a subset of all events by specifying `offset`, or use `0L` to retrieve all
+   * events with a given tag. The `offset` corresponds to a timesamp of the events. Note that the
+   * corresponding offset of each event is provided in the `akka.persistence.query.EventEnvelope`,
+   * which makes it possible to resume the stream at a later point from a given offset.
+   * The `offset` is inclusive, i.e. the events with the exact same timestamp will be included
+   * in the returned stream.
+   *
+   * There is a variant of this query with a time based UUID as offset. That query is better
+   * for cases where you want to resume the stream from an exact point without receiving
+   * duplicate events for the same timestamp.
```
Comment (patriknw) [inline src/main/scala/akka/persistence/cassandra/query/scaladsl/CassandraReadJournal.scala]: This and perhaps some more of this documentation needs an update.

File: src/main/scala/akka/persistence/cassandra/query/scaladsl/CassandraReadJournal.scala [hunk 2]
```diff
@@ -468,6 +555,34 @@ class CassandraReadJournal(system: ExtendedActorSystem, config: Config)
...
+
+  private[this] def internalUuidToOffset(uuid: UUID): Offset =
+    if (uuid == firstOffset) NoOffset
+    else TimeBasedUUID(uuid)
+
+  private[this] def offsetToInternalOffset(offset: Offset): UUID =
+    offset match {
+      case TimeBasedUUID(uuid) => uuid
+      case NoOffset            => firstOffset
+      case unsupported =>
```
Comment (patriknw) [inline src/main/scala/akka/persistence/cassandra/query/scaladsl/CassandraReadJournal.scala]: shouldn't `Sequence` be supported? it corresponds to the previous `Long` variant

File: src/main/scala/akka/persistence/cassandra/query/scaladsl/CassandraReadJournal.scala [hunk 3]
```diff
@@ -283,26 +369,45 @@ class CassandraReadJournal(system: ExtendedActorSystem, config: Config)
...
+  @deprecated("use eventsByTag(String, Offset) instead", "0.20")
   def eventsByTag(tag: String, offset: UUID): Source[UUIDEventEnvelope, NotUsed] =
+    eventsByTag(tag, internalUuidToOffset(offset)).map(eventEnvelope2toUuidEnvelope)
+
+  /**
+   * Same type of query as `eventsByTag` but the event stream
+   * is completed immediately when it reaches the end of the "result set". Events that are
+   * stored after the query is completed are not included in the event stream.
+   *
+   * Use [[#firstOffset]] when you want all events from the beginning of time.
```
Comment (patriknw) [inline src/main/scala/akka/persistence/cassandra/query/scaladsl/CassandraReadJournal.scala]: update

File: src/main/scala/akka/persistence/cassandra/query/scaladsl/CassandraReadJournal.scala (hunk 2 above)
Comment (johanandren) [inline src/main/scala/akka/persistence/cassandra/query/scaladsl/CassandraReadJournal.scala]: Just to make sure I'm clear on the semantics here, the user makes a query with `Sequence(value)` (where value is a timestamp and not a sequence number), the returned envelopes then have the `TimeBasedUUID` for each event or do we do a lossy (loosing the rest of the uuid except for the timestamp) transformation back based on the original query type?

Also, isn't this terribly confusing compared to only accepting `TimeBasedUUID` and the user can create the uuid using `UUID.fromTime(timestamp)` themselves?

Comment (patriknw) [inline src/main/scala/akka/persistence/cassandra/query/scaladsl/CassandraReadJournal.scala]: Yeah, I think I mixed up things. Sequence should not be used for Timestamp (old Long methods here). Timestamp should perhaps be another Offset type? You can start with only TimeBasedUUID until we have thought this through

Comment (johanandren) [inline src/main/scala/akka/persistence/cassandra/query/scaladsl/CassandraReadJournal.scala]: Ok, thanks for clarifying.

File: src/main/scala/akka/persistence/cassandra/query/scaladsl/CassandraReadJournal.scala [hunk 4]
```diff
@@ -468,6 +550,35 @@ class CassandraReadJournal(system: ExtendedActorSystem, config: Config)
...
+    )
+
+  private[this] def internalUuidToOffset(uuid: UUID): Offset =
+    if (uuid == firstOffset) NoOffset
+    else TimeBasedUUID(uuid)
+
+  private[this] def offsetToInternalOffset(offset: Offset): UUID =
+    offset match {
+      case TimeBasedUUID(uuid) => uuid
+      case Sequence(timestamp) => offsetUuid(timestamp)
```
Comment (patriknw) [inline src/main/scala/akka/persistence/cassandra/query/scaladsl/CassandraReadJournal.scala]: Didn't we conclude that Sequence isn't the same as a wall clock timestamp?
After your comment about that I was convinced that we should only support the `TimeBasedUUID`.

We should also add `def timeBasedUUIDFrom(timestamp: Long): TimeBasedUUID` and mention that in the docs of `eventsByTag`. Note that `def offsetUuid(timestamp: Long): UUID` already exists.

Comment (johanandren) [inline src/main/scala/akka/persistence/cassandra/query/scaladsl/CassandraReadJournal.scala]: You are correct, I think this was a leftover.

The `timeBasedUUIDFrom` as a public method?

Comment (patriknw) [inline src/main/scala/akka/persistence/cassandra/query/scaladsl/CassandraReadJournal.scala]: yes, that was my thinking.
if you want to start from a specific time but don't care (or know) the exact uuid offset

**Pull request merged**